woocommerce/woocommerce
# Add the wc_allow_changing_orders_storage_while_sync_is_pending filter

### Changes proposed in this Pull Request:

For HPOS the idea is to disallow changing the authoritative data source for orders (between the new tables and the posts table) while there are orders pending synchronization, since doing that could lead to order data corruption/loss. However, while this was indeed disallowed in the features UI, it was still possible to manually change the value of the corresponding option (via `wp option` command); there was code in place to prevent this (by throwing an exception) but it had been commented out to ease
testing during the development of the feature.

This pull request restores that commented out code, so in principle, trying to change value of the option while there are orders pending synchronization will throw an exception. **But** a new filter is introduced, `wc_allow_changing_orders_storage_while_sync_is_pending`, to allow bypassing this behavior: the filter returns `false` by default, but if it returns `true`, changing the option will never throw an exception. This is needed for a few of the existing unit tests,
but otherwise it should never be done (and the filter is documented as such).

Additionally, if the filter returns `true` it will be possible to change the value of the option in the features UI; but a warning
message will be shown in red if there are orders pending sync (see screenshots below, without and with the filter).

![image](https://github.com/woocommerce/woocommerce/assets/937723/81b31f10-c51b-4254-b79c-bc3f1a52d440)

![image](https://github.com/woocommerce/woocommerce/assets/937723/9bc5086c-42c8-4a7f-95e8-f4fcad896a7b)


### How to test the changes in this Pull Request:



1. Start going to WooCommerce - Settings - Advanced - Features, select HPOS in "Data storage for orders" and disable synchronization.
2. Verify that you can change the data storage setting from the settings UI and no errors pop up, also there are no warning messages next to the setting (leave the setting as HPOS before continuing).
3. Run the following command from the command line: `wp option set woocommerce_custom_orders_table_enabled no`. It should work with no errors. Run the command again with `yes` instead of `no`.
4. Create an order or edit an existing order (changing its status is enough).
5. Reload the settings page, verify that the "Data storage for orders" radio buttons are disabled and you see the "There's 1 order pendync sync. You can switch data storage..." message in gray.
6. Try again `wp option set woocommerce_custom_orders_table_enabled no` from the command line. It should throw an exception.
7. Run `wp wc cot sync` from the command line so there aren't orders pending sync anymore.
8. Add the following code at the end of `woocommerce.php` or using a snippets plugin: `add_filter('wc_allow_changing_orders_storage_while_sync_is_pending', '__return_true');`
9. Reload the settings UI, repeat steps 2 and 3, verify that nothing has changed.
10. Repeat step 4 and reload the settings page, this time the radio buttons are enabled and the message is "There's 1 order pending sync. Switching data storage while sync is incomplete is dangerous..." in red.
11. Verify that indeed you can change the value of "Data storage for orders". Leave the setting as HPOS before continuing.
12. Repeat step 6, this time it works and no exception is thrown.

### Changelog entry




<details>

#### Significance




#### Type




#### Message

Throw exception if order data storage change is attempted while there are orders pending sync, and add the wc_allow_changing_orders_storage_while_sync_is_pending filter to bypass this

#### Comment

</details>
